### PR TITLE
Introducing ruff as a linter to replace pylint, flake8 and isort

### PR DESCRIPTION
**Pull request by noklam**
Signed-off-by: Nok <nok.lam.chan@quantumblack.com>

> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description

The goal for this is to use `ruff` to replace `pylint`, we should aim at compatible (99.9%) for this scope. We can decide later rather we should enable extra rule set.

`ruff` is much fast than `pylint`, which kills dev productivity for trivial changes. `black` will be kept because `ruff` is not a formatter although it can flag issues with formatting.
## Development notes

- Update `linting.md`
- Remove `isort`,`flake8`,`pylint` from our CI (noted the dependencies are still here because we only deprecate `kedro lint` in 0.19.0`
## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Comment by noklam** [comment]
Playing around with it, it's amazingly fast. I get confused a little bit with max-length. It seems that it can only warn the line is too long but couldn't fix it. 

So I configure it to ignore E501(line-length), and just run black over it instead. There are subtle difference so they are not 1:1 mapping, but I guess `black` is doing a good job there and I am not worry about it.

**Comment by noklam** [comment]
To test this you can still uncomment the existing setup. They are mostly identical but with some slight difference as `ruff` is not a 100% replacement to the existing tools.

**Comment by datajoely** [comment]
❤️

**Comment by astrojuanlu** [comment]
This still calls pylint right? Maybe we can remove it from CI to see the difference in execution times

**Comment by noklam** [comment]
@astrojuanlu Ruff run almost instantly compare to pylint :)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -32,14 +61,15 @@ repos:
           - id: pyupgrade
             args: [--py36-plus]
 
+    - repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (astrojuanlu) [inline .pre-commit-config.yaml:64]: I think `ruff` is running twice 😄

File: test_requirements.txt [hunk 2]
```diff
@@ -20,7 +20,6 @@ holoviews~=1.13.0
 import-linter[toml]==1.8.0
 ipython>=7.31.1, <8.0; python_version < '3.8'
 ipython~=8.10; python_version >= '3.8'
-isort~=5.0
 Jinja2<3.1.0
 joblib>=0.14
 jupyterlab_server>=2.11.1, <2.16.0 # 2.16.0 requires importlib_metedata >= 4.8.3 which conflicts with flake8 requirement
```
Comment (astrojuanlu) [inline test_requirements.txt:25]: Does this mean that we don't need to cap `jupyterlab_server` anymore?

**Comment by astrojuanlu** [comment]
- Before 🐢 https://app.circleci.com/pipelines/github/kedro-org/kedro/24005/workflows/11d8de9f-5bed-4a12-a158-7653018def2c/jobs/277033 5m 48s (total), 1m 29s (Run linters)
- After 🐇 https://app.circleci.com/pipelines/github/kedro-org/kedro/24027/workflows/d93195aa-83b6-4872-8c0f-9505724e0f08/jobs/277528 4m 43s (total), 47s (Run linters)

Let's get this merged!

**Review by astrojuanlu — APPROVED** [review]
LGTM! We'd need to make a couple of updates to the dev docs that still mention flake8 and friends, but I'd say let's leave those to #2376, wdyt @stichbury?

**Comment by stichbury** [comment]
> LGTM! We'd need to make a couple of updates to the dev docs that still mention flake8 and friends, but I'd say let's leave those to #2376, wdyt @stichbury?

I'd prefer to add those changes to this ticket if possible please, because I think it's more relevant here and best to pair them in the same PR. The other ticket is more about contributor docs as I see it, rather than tooling, although it may touch there.

@noklam are you OK to make the changes that @astrojuanlu is suggesting? I can review them ofc.

**Comment by noklam** [comment]
> * Before 🐢 https://app.circleci.com/pipelines/github/kedro-org/kedro/24005/workflows/11d8de9f-5bed-4a12-a158-7653018def2c/jobs/277033 5m 48s (total), 1m 29s (Run linters)
> * After 🐇 https://app.circleci.com/pipelines/github/kedro-org/kedro/24027/workflows/d93195aa-83b6-4872-8c0f-9505724e0f08/jobs/277528 4m 43s (total), 47s (Run linters)
> 
> Let's get this merged! @astrojuanlu 

This includes the time to install the linter. It will be much faster during development, can't wait to show it in Sprint Demo :)

**Comment by noklam** [comment]
> I'd prefer to add those changes to this ticket if possible please, because I think it's more relevant here and best to pair them in the same PR. The other ticket is more about contributor docs as I see it, rather than tooling, although it may touch there.
> 
> @noklam are you OK to make the changes that @astrojuanlu is suggesting? I can review them ofc.

👍🏼 I will do this

**Comment by noklam** [comment]
Remove `setup.cfg` from docs and starters

File: docs/source/development/commands_reference.md [hunk 3]
```diff
@@ -402,7 +402,7 @@ The `build-docs` command builds [project documentation](../tutorial/package_a_pr
 #### Lint your project
 
 ```{note}
-_This command will be deprecated from Kedro version 0.19.0._
+_This command will be deprecated from Kedro version 0.19.0._. We still recommend to (../development/linting.md) and you can find more help here
```
Comment (merelcht) [inline docs/source/development/commands_reference.md:405]: Is the word "here" supposed to link somewhere or do you mean users can find more on the linting page? In that case I'd change this to: "and you can find more information about linting on that page."

File: docs/source/development/linting.md [hunk 4]
```diff
@@ -10,25 +10,23 @@ As a project grows and goes through various stages of development it becomes imp
...
+[`black`](https://github.com/psf/black), [`ruff`](https://github.com/PyCQA/flake8)
 - **`black`** is a [PEP 8](https://peps.python.org/pep-0008/) compliant opinionated Python code formatter. `black` can
 check for styling inconsistencies and reformat your files in place.
 [You can read more in the `black` documentation](https://black.readthedocs.io/en/stable/).
-- **`flake8`** is a wrapper around [`pep8`](https://pypi.org/project/pep8/),
-[`pyflakes`](https://pypi.org/project/pyflakes/), and [`mccabe`](https://pypi.org/project/mccabe/) which can lint code and format it with respect to [PEP 8](https://peps.python.org/pep-0008/),
-and check the [cyclomatic complexity](https://www.ibm.com/docs/en/raa/6.1?topic=metrics-cyclomatic-complexity) of your code base.
-[You can read more in the `flake8` documentation](https://flake8.pycqa.org/en/latest/).
-- **`isort`** is a Python library used to reformat code by sorting imports alphabetically and automatically separating them into sections by
+- **`ruff`** is a fast linter that replace `flake8`, `isort` and `pylint`.
```
Comment (merelcht) [inline docs/source/development/linting.md:17]:
```suggestion
- **`ruff`** is a fast linter that replaces `flake8`, `isort` and `pylint`.
```

File: docs/source/development/linting.md [hunk 5]
```diff
@@ -10,25 +10,23 @@ As a project grows and goes through various stages of development it becomes imp
...
+  - It helps your code compliant to [`pep8`](https://pypi.org/project/pep8/).
```
Comment (merelcht) [inline docs/source/development/linting.md:18]:
```suggestion
  - It helps to make your code compliant to [`pep8`](https://pypi.org/project/pep8/).
```

File: docs/source/development/linting.md [hunk 6]
```diff
@@ -10,25 +10,23 @@ As a project grows and goes through various stages of development it becomes imp
...
+  - It reformat code by sorting imports alphabetically and automatically separating them into sections by
```
Comment (merelcht) [inline docs/source/development/linting.md:19]:
```suggestion
  - It reformats code by sorting imports alphabetically and automatically separating them into sections by
```

File: pyproject.toml [hunk 7]
```diff
@@ -175,3 +176,20 @@ ignore_imports = [
     "kedro.framework.context.context -> kedro.config",
     "kedro.framework.session.session -> kedro.config"
 ]
+
+[tool.ruff]
+line-length = 88
+fix = false
+show-fixes = true
+# select = ["A", "B", "C", "E", "F", "G", "I", "N", "Q", "S", "T", "W", "ARG", "BLE", "COM", "DJ", "DTZ", "EM", "ERA", "EXE", "FBT", "ICN", "INP", "ISC", "NPY", "PD", "PGH", "PIE", "PL", "PT", "PTH", "PYI", "RET", "RSE", "RUF", "SIM", "SLF", "TCH", "TID", "TRY", "UP", "YTT"]
```
Comment (merelcht) [inline pyproject.toml:184]: What does all of this do? And does it need to stay as a comment?

**Review by merelcht — COMMENTED** [review]
This looks awesome! Thanks for introducing `ruff` @noklam 😄 

I left some small comments and questions. I think the main point I want to echo is the @astrojuanlu left about `ruff` appearing twice in the `pre-commit` file. Is that on purpose?

**Comment by noklam** [comment]
Requesting @stichbury for doc review. I still need to fix a lot of linting error. It's not hard but takes time, will resume when I am back.

**Comment by noklam** [comment]
There are many linting error, many of them are not difference between `ruff` and `pylint` but rather we are using old pylint version.

For the changes that are obvious and meaningful, I have applied the changes and suppressed the rest.

For example, this is a good fix hinted by `ruff` [e0275f0](https://github.com/kedro-org/kedro/pull/2634/commits/e0275f0dc3b7a7e327ec6408b5776ef22829bf2f).

File: kedro/extras/datasets/geopandas/geojson_dataset.py [hunk 8]
```diff
@@ -52,8 +52,7 @@ class GeoJSONDataSet(
     DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
     DEFAULT_SAVE_ARGS = {"driver": "GeoJSON"}
 
-    # pylint: disable=too-many-arguments
-    def __init__(
+    def __init__(  # noqa: too-many-arguments
```
Comment (sbrugman) [inline kedro/extras/datasets/geopandas/geojson_dataset.py:55]: Consider a global ignore of this rule

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -19,27 +19,58 @@ repos:
...
+      # Ruff version.
+      rev: v0.0.270
       hooks:
-          - id: pyupgrade
-            args: [--py36-plus]
+        - id: ruff
+          name: "ruff on tests/ and docs/"
+          # PLR2004: Magic value used
+          # PLR0913: Too many arguments
+          args: ["--fix", "--exit-non-zero-on-fix", "--ignore=PLR2004,PLR0913"]
```
[the remark this answers is not in the document]
Comment (noklam) [reply .pre-commit-config.yaml:68]: @sbrugman Could you share a link? Not sure if I missed anything, I checked their config but it does not use any arguments.

https://github.com/ing-bank/popmon/blob/ac79d212a519368d01525950142e0a282f5287c3/.pre-commit-config.yaml#L6-L10

File: docs/source/development/linting.md [hunk 10]
```diff
@@ -10,25 +10,23 @@ As a project grows and goes through various stages of development it becomes imp
 
 ## Set up Python tools
 There are a variety of Python tools available to use with your Kedro projects. This guide shows you how to use
-[`black`](https://github.com/psf/black), [`flake8`](https://github.com/PyCQA/flake8), and
-[`isort`](https://github.com/PyCQA/isort) to format and lint your Kedro projects.
+[`black`](https://github.com/psf/black), [`ruff`](https://github.com/PyCQA/flake8).
```
[the remark this answers is not in the document]
Comment (noklam) [reply docs/source/development/linting.md:13]: Good spot, typo.

File: .pre-commit-config.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sbrugman) [reply .pre-commit-config.yaml]: See the section [tool.ruff.per-file-ignores] in pyproject.toml

**Review by merelcht — APPROVED** [review]
Great work @noklam ⭐  I went through most of the changed files and can see that the changes minimal and mostly just changing `# pylint: disable` to `# noqa`.

**Comment by noklam** [comment]
I don't want this PR stuck forever, since this is already very useful to speed up dev.

Follows up:
- See the section [tool.ruff.per-file-ignores] in pyproject.toml
- Replace `bandit` with ruff

**Comment by noklam** [comment]
No, I will save this for future PRs.

The lining is still failing for weird reason, I cannot reproduce it anywhere so I may have to ssh into it.

Other than that it's block by the E2e test.

**Comment by noklam** [comment]
The unsort import error is mysteriou, I cannot reproduce it anywhere. It's not important at all so I silent this error for the `micropkg.py`.

**Comment by noklam** [comment]
Linting now all pass, only blocked by #2816, enable auto-merge.

**Pull request merged**